### DISCUSSION TITLE
Visualizing a Replayed Code Execution

### DISCUSSION THREAD

**Discussion by cretz** [code-hike/codehike#92] [Show and tell]
I work at https://temporal.io/ which does workflows as code, and as a side project I've been toying with how to visualize past runs. Temporal itself supports taking a replay and running against the previous code which is what I do at https://github.com/cretz/temporal-debug-go with a debugger to capture some tracing info.

I added a Code Hike based HTML exporter using a Scrollycoding MDX. It uses Go to generate a dynamic MDX that is rendered to a static site with a Next.js project (see Go code [here](https://github.com/cretz/temporal-debug-go/blob/main/tracer/html_annotated.go) that creates dynamic files underneath [this dir](https://github.com/cretz/temporal-debug-go/tree/main/tracer/html_annotated_proj)).

[See an example of the output here](https://cretz.github.io/temporal-debug-go/examples/cancellation/html-annotated/).

In this particular example, there is way less code than events, but other run visualizations have much more code and fewer events. But hopefully that example shows a neat use case for Code Hike. It unfortunately has some scrolling jank (I wish I could turn scrolling off, but this is still a better view than anything I could get with Spotlight).

I could see a more generic version of such a thing backed by a trace from https://github.com/rr-debugger/rr.

I eagerly look forward to #81 and/or any other visualization technique or improvements on the existing ones to show replayed code interspersed with contextual info (in my case, events from the server).

**Comment by pomber** [comment]
So cool you are experimenting with this. That's the idea of #81, but I'm waiting for a few use cases (like yours) so I can generalize the API and the output.  

Do you mind sharing what do you think would be the ideal UI of the output?

**Reply by cretz** [reply]
> Do you mind sharing what do you think would be the ideal UI of the output?

I am admittedly very weak with visualizations and UI. I think what is there is fine now ignoring scrolling. Ideally I would use the Spotlight but I was struggling with some details of it (I can't recall what they were off hand).

I will say one of my struggles of using the system, due to not being well-versed in modern web tech, is that converting this MDX to a static HTML was expensive via Next.js. I have a set of code files and steps that change which parts of them are visible and I want that to appear in the page, but Next takes many many seconds to generate an optimized build of a static site. Do you have any other MDX-to-HTML output suggestions?

**Reply by Aiyush-G** [reply]
I also noticed when on mobile, the design isn't responsive and unusable, it would be good if it a) became responsive or b) showed a stripped down version?

Following on from what @cretz has said, I think that it would be awesome if there was a web API in place that could be hosted locally that would take in the Markdown file and then spit out static HTML, there could be options for the styling etc... separate style sheets and custom behaviours. This would also make Code Hike more accessible in the whole developer ecosystem as an application that wasn't built with react could also make use of these static HTML files, refer to #90 . I'm not sure if this is doable, what do you think?

EDIT: Clarify API to web API

**Reply by pomber** [reply]
Maybe I'm not understanding your use case, but doesn't it make more sense to have something like this:

![image](https://user-images.githubusercontent.com/1911623/145589279-9e5cefd2-ff7c-47ee-ae49-c59e7ff683a0.png)

A separate view for the code and the output, and a slider or something to change the steps?

**Reply by cretz** [reply]
That would work quite well, yes! But I'm ok with only one code area too. Also, while the example I have doesn't show it, the code portion can span multiple files. I was also wanting to do hover-over or another approach to display the contents of local variables too, so I wonder if that "history.json" area there can serve as more of "contextual state area". Then though you are basically rewriting a debugger UI. One benefit of a ever-present left nav portion instead of a slider is the ability to see all the steps at once. I wonder if a vertical slider is a thing (not to be confused with the scrollbar; I personally don't like hijacking a user's scroll).

My limited UI skills often have me reaching for something that already exists, heh.

**Reply by nadilas** [reply]
I quite like the "ever-present" navigation sidebar at least for the use case of temporal that makes the context of the code clearer.
To me the slider (it being horizontal) is counterintuitive when thinking of flow of events or code, I'd much rather have a vertical slider in the middle as a sort of central function.

> I was also wanting to do hover-over or another approach to display the contents of local variables too, so I wonder if that "history.json" area there can serve as more of "contextual state area".

Since I am quite used to values being inline with the option to expand, it'd be nice to consider if we could do: https://www.jetbrains.com/help/rider/Inspecting_Variables.html#inline-with-code

**Reply by pomber** [reply]
For the vertical navigation sidebar, I think `CH.Spotlight` should work better than `CH.Scrollycoding`. If you add some custom CSS it should look at least decent.

For the inline values and tooltips, it should be possible to add Annotations. See the [label annotation](https://codehike.org/demo/comment-annotations) and the [tooltip custom annotation](https://codehike.org/demo/custom-annotations).

**Comment by pomber** [comment]
I created #93 for the animation jank for big files

**Reply by cretz** [reply]
I actually like that the full file is available for those wanting to scroll (well, actually not even the full file is available). My issue is that, for instance, it starts on the middle event when you get to the webpage, it moves when you scroll in potentially unexpected ways, etc. I suppose my use case is more fitting for a Spotlight (I don't care about scrolling). When I get back to the project, I'll try to enumerate my struggles with Spotlight.

**Reply by nadilas** [reply]
I just looked at cretz's example and found that on a much larger screen the scroll behavior selects the 8th nav item even though I'm at the top. Scroll down and scrolling back up, I can't get higher than the 9th nav item. Not by scrolling at least. Selecting the first item and scrolling down a little, jumps immediately to 10-13th.
